src/hotspot/os/linux/os_linux.cpp

```diff
@@ -2888,6 +2893,32 @@ void os::pd_free_memory(char *addr, size_t bytes, size_t alignment_hint) {
   }
 }
 
+static void warn_fail_pretouch_memory(void *first, void *last, size_t page_size,
+                                      int err) {
+  warning("INFO: os::pd_pretouch_memory(" PTR_FORMAT ", " PTR_FORMAT ", "
+          SIZE_FORMAT ") failed; error='%s' (errno=%d)",
+          p2i(first), p2i(last), page_size,
+          os::strerror(err), err);
+}
+
+void os::pd_pretouch_memory(void *first, void *last, size_t page_size) {
+  size_t len = static_cast<char *>(last) - static_cast<char *>(first) + page_size;
+  // Use madvise to pretouch on Linux first, and fallback to the common method
+  // if unsupported. THP can form right after madvise rather than being
+  // assembled later.
+  if (::madvise(first, len, MADV_POPULATE_WRITE) == -1) {
+    int err = errno;
+    if (err == EINVAL) { // Not supported
```

Comment:
Would be nice to avoid repeated syscalls to madvise if this fails once; no reason to try again, then.

Reply:
I tested the performance of this patch on kernel 4.18 and 5.13, and found the repeat calls have no impact. So I would not change anything about this.

Reply:
It seems that a thread-local variable is cheap for recording whether madvise once failed, but os::Linux does not have a thread-local variable yet. Is it okay to have a thread-local variable for this?